promql/info.go

```diff
@@ -0,0 +1,467 @@
+	annots.Merge(ws)
+	if err != nil {
+		annots.Add(err)
+		return nil, annots
+	}
+
+	res, ws := ev.combineWithInfoSeries(ctx, mat, infoSeries, ignoreSeries, dataLabelMatchers)
+	annots.Merge(ws)
+	return res, annots
+}
+
+// fetchInfoSeries fetches info series given ev.selectHints and matching identifying labels in mat.
+// Series in ignoreSeries are not fetched.
+func (ev *evaluator) fetchInfoSeries(ctx context.Context, mat Matrix, ignoreSeries map[int]struct{}, dataLabelMatchers map[string][]*labels.Matcher) (Matrix, annotations.Annotations, error) {
+	if ev.selectHints == nil {
+		// ev.selectHints should have been set.
+		var annots annotations.Annotations
+		annots.Add(fmt.Errorf("ev.selectHints not set"))
+		return nil, annots, nil
+	}
+
+	// A map of values for all identifying labels we are interested in.
+	idLblValues := map[string]map[string]struct{}{}
+	for i, s := range mat {
+		if _, exists := ignoreSeries[i]; exists {
+			continue
+		}
+
+		// Register relevant values per identifying label for this series.
+		lblMap := s.Metric.Map()
+		for _, l := range identifyingLabels {
+			val := lblMap[l]
+			if val == "" {
+				continue
+			}
+
+			if idLblValues[l] == nil {
+				idLblValues[l] = map[string]struct{}{}
+			}
+			idLblValues[l][val] = struct{}{}
+		}
+	}
+	if len(idLblValues) == 0 {
+		return nil, nil, nil
+	}
+
+	// Generate regexps for every interesting value per identifying label.
+	var sb strings.Builder
+	idLblRegexps := make(map[string]string, len(idLblValues))
+	for name, vals := range idLblValues {
+		sb.Reset()
+		i := 0
+		for v := range vals {
+			if i > 0 {
+				sb.WriteRune('|')
+			}
+			sb.WriteString(regexp.QuoteMeta(v))
+			i++
+		}
+		idLblRegexps[name] = sb.String()
+	}
+
+	var infoLabelMatchers []*labels.Matcher
+	for name, re := range idLblRegexps {
+		infoLabelMatchers = append(infoLabelMatchers, labels.MustNewMatcher(labels.MatchRegexp, name, re))
+	}
+	var nameMatcher *labels.Matcher
+	for name, ms := range dataLabelMatchers {
+		for i, m := range ms {
+			if m.Name == labels.MetricName {
+				nameMatcher = m
+				ms = slices.Delete(ms, i, i+1)
+			}
+			infoLabelMatchers = append(infoLabelMatchers, m)
+		}
+		if len(ms) > 0 {
+			dataLabelMatchers[name] = ms
+		} else {
+			delete(dataLabelMatchers, name)
+		}
+	}
+	if nameMatcher == nil {
+		// Default to using the target_info metric.
+		infoLabelMatchers = append([]*labels.Matcher{labels.MustNewMatcher(labels.MatchRegexp, labels.MetricName, targetInfo)}, infoLabelMatchers...)
+	}
+
+	infoIt := ev.querier.Select(ctx, false, ev.selectHints, infoLabelMatchers...)
+	annots := infoIt.Warnings()
+	if infoIt.Err() != nil {
+		return nil, annots, infoIt.Err()
+	}
+	var infoSeries []storage.Series
+	for infoIt.Next() {
+		select {
+		case <-ctx.Done():
+			return nil, nil, ctx.Err()
+		default:
+		}
+		infoSeries = append(infoSeries, infoIt.At())
+	}
+	infoMat := ev.expandSeriesToMatrix(ctx, infoSeries, 0, true)
```

Comment:
We need to thread offset of the labelselector into this call here. Otherwise it selects the wrong samples.

Reply:
We should take the offset of the "LabelSelector" argument of the info function and just pass it down.

Reply:
By "LabelSelector", you mean the second argument to `info`? What would lead to that having an offset? Can you provide example PromQL?

Reply:
i.e. `info(X offset 1m, {job=~".+"} offset 1m)`

Reply:
What are the semantics of `{job=~".+"} offset 1m` here?

Reply:
It would be evaluated when joining the info series to the result of the evaluation of the expression in the first argument.

Reply:
I don't think it should be possible to specify offset for the second argument quite simply.

Reply:
The second argument should only carry information about data label matchers.

Reply:
That can be checked at parse time.